messaging/messaging_test.go

```diff
@@ -0,0 +1,829 @@
+				TTL: &invalidTTL,
+			},
+			Topic: "topic",
+		},
+		want: "ttl duration must not be negative",
+	},
+	{
+		name: "InvalidAndroidPriority",
+		req: &Message{
+			Android: &AndroidConfig{
+				Priority: "not normal",
+			},
+			Topic: "topic",
+		},
+		want: "priority must be 'normal' or 'high'",
+	},
+	{
+		name: "InvalidAndroidColor1",
+		req: &Message{
+			Android: &AndroidConfig{
+				Notification: &AndroidNotification{
+					Color: "112233",
+				},
+			},
+			Topic: "topic",
+		},
+		want: "color must be in the #RRGGBB form",
+	},
+	{
+		name: "InvalidAndroidColor2",
+		req: &Message{
+			Android: &AndroidConfig{
+				Notification: &AndroidNotification{
+					Color: "#112233X",
+				},
+			},
+			Topic: "topic",
+		},
+		want: "color must be in the #RRGGBB form",
+	},
+	{
+		name: "InvalidAndroidTitleLocArgs",
+		req: &Message{
+			Android: &AndroidConfig{
+				Notification: &AndroidNotification{
+					TitleLocArgs: []string{"a1"},
+				},
+			},
+			Topic: "topic",
+		},
+		want: "titleLocKey is required when specifying titleLocArgs",
+	},
+	{
+		name: "InvalidAndroidBodyLocArgs",
+		req: &Message{
+			Android: &AndroidConfig{
+				Notification: &AndroidNotification{
+					BodyLocArgs: []string{"a1"},
+				},
+			},
+			Topic: "topic",
+		},
+		want: "bodyLocKey is required when specifying bodyLocArgs",
+	},
+	{
+		name: "APNSMultipleAlerts",
+		req: &Message{
+			APNS: &APNSConfig{
+				Payload: &APNSPayload{
+					Aps: &Aps{
+						Alert:       &ApsAlert{},
+						AlertString: "alert",
+					},
+				},
+			},
+			Topic: "topic",
+		},
+		want: "multiple alert specifications",
+	},
+	{
+		name: "InvalidAPNSTitleLocArgs",
+		req: &Message{
+			APNS: &APNSConfig{
+				Payload: &APNSPayload{
+					Aps: &Aps{
+						Alert: &ApsAlert{
+							TitleLocArgs: []string{"a1"},
+						},
+					},
+				},
+			},
+			Topic: "topic",
+		},
+		want: "titleLocKey is required when specifying titleLocArgs",
+	},
+	{
+		name: "InvalidAPNSLocArgs",
+		req: &Message{
+			APNS: &APNSConfig{
+				Payload: &APNSPayload{
+					Aps: &Aps{
+						Alert: &ApsAlert{
+							LocArgs: []string{"a1"},
+						},
+					},
+				},
+			},
+			Topic: "topic",
+		},
+		want: "locKey is required when specifying locArgs",
+	},
+}
+
+var invalidTopicMgtArgs = []struct {
+	name   string
+	tokens []string
+	topic  string
+	want   string
+}{
+	{
+		name: "NoTokensAndTopic",
+		want: "no tokens specified",
+	},
+	{
+		name:   "NoTopic",
+		tokens: []string{"token1"},
+		want:   "topic name not specified",
+	},
+	{
+		name:   "InvalidTopicName",
+		tokens: []string{"token1"},
+		topic:  "foo*bar",
+		want:   "invalid topic name: \"foo*bar\"",
+	},
+	{
+		name:   "TooManyTokens",
+		tokens: strings.Split("a"+strings.Repeat(",a", 1000), ","),
+		topic:  "topic",
+		want:   "tokens list must not contain more than 1000 items",
+	},
+	{
+		name:   "EmptyToken",
+		tokens: []string{"foo", ""},
+		topic:  "topic",
+		want:   "tokens list must not contain empty strings",
```

Comment:
I've been told previously by other Go readability reviewers that string comparison on error messages is brittle and thus discouraged, but I'm having trouble finding a referenceable link that addresses this issue and what they suggest otherwise.

Reply:
I agree in general. However, at least in this case we are testing for the errors produced by our own code.

Reply:
Good point. SG.